src/lavinmq/exchange/headers.cr

```diff
@@ -13,44 +13,55 @@ module LavinMQ
       super
     end
 
-    def bind(destination : Queue, routing_key, headers)
-      validate!(headers)
-      args = headers ? @arguments.clone.merge!(headers) : @arguments
-      ret = @queue_bindings[{routing_key, args}].add? destination
-      after_bind(destination, routing_key, headers)
-      ret
+    def type : String
+      "headers"
     end
 
-    def bind(destination : Exchange, routing_key, headers)
-      validate!(headers)
-      args = headers ? @arguments.clone.merge!(headers) : @arguments
-      ret = @exchange_bindings[{routing_key, args}].add? destination
-      after_bind(destination, routing_key, headers)
-      ret
+    def bindings_details : Iterator(BindingDetails)
+      @bindings.each.flat_map do |args, ds|
+        ds.map do |d|
+          binding_key = BindingKey.new("", args)
+          BindingDetails.new(name, vhost.name, binding_key, d)
+        end
+      end
     end
 
-    def unbind(destination : Queue, routing_key, headers)
+    def bind(destination : Destination, routing_key, headers)
+      validate!(headers)
       args = headers ? @arguments.clone.merge!(headers) : @arguments
-      ret = @queue_bindings[{routing_key, args}].delete destination
-      after_unbind(destination, routing_key, headers)
-      ret
+      if @bindings[args].add? destination
+        binding_key = BindingKey.new(routing_key, args)
+        data = BindingDetails.new(name, vhost.name, binding_key, destination)
+        notify_observers(ExchangeEvent::Bind, data)
+        true
+      else
+        false
+      end
     end
 
-    def unbind(destination : Exchange, routing_key, headers)
+    def unbind(destination : Destination, routing_key, headers)
       args = headers ? @arguments.clone.merge!(headers) : @arguments
-      ret = @exchange_bindings[{routing_key, args}].delete destination
-      after_unbind(destination, routing_key, headers)
-      ret
-    end
+      bds = @bindings[args]
+      if bds.delete(destination)
+        @bindings.delete(routing_key) if bds.empty?
+
+        binding_key = BindingKey.new(routing_key, args)
+        data = BindingDetails.new(name, vhost.name, binding_key, destination)
+        notify_observers(ExchangeEvent::Unbind, data)
 
-    def do_queue_matches(routing_key, headers = nil, & : Queue ->)
-      matches(@queue_bindings, routing_key, headers) do |destination|
-        yield destination.as(Queue)
+        delete if @auto_delete && @bindings.each_value.all?(&.empty?)
+        true
+      else
+        false
```

Comment:
```suggestion
```
(also remove `end` on next line, but couldn't get it in the suggestion because it's not part of a change?)